server/rest/src/main/java/org/infinispan/rest/stream/CacheEntryStreamProcessor.java

```diff
@@ -0,0 +1,169 @@
+package org.infinispan.rest.stream;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+import org.infinispan.commons.dataconversion.internal.Json;
+import org.infinispan.container.entries.CacheEntry;
+import org.infinispan.container.entries.InternalCacheEntry;
+import org.reactivestreams.Publisher;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.channel.ChannelHandlerContext;
+
+/**
+ * A {@link CacheChunkedStream} that reads {@link Map.Entry} and produces a JSON output.
+ * For example:
+ * <p>
+ * <pre>{@code
+ * [
+ *   {"key":1,"value":"value","timeToLiveSeconds": -1, "maxIdleTimeSeconds": -1, "created": -1, "lastUsed": -1, "expireTime": -1},
+ *   {"key":2,"value":"value2","timeToLiveSeconds": -1, "maxIdleTimeSeconds": -1, "created": -1, "lastUsed": -1, "expireTime": -1}
+ * ]
+ * }</pre>
+ *
+ *
+ * @since 12.0
+ */
+public class CacheEntryStreamProcessor extends CacheChunkedStream<CacheEntry<?, ?>> {
+   private final boolean keysAreJson;
+   private final boolean valuesAreJson;
+
+   private static final byte[] KEY_LABEL = "\"key\":".getBytes();
+   private static final byte[] VALUE_LABEL = "\"value\":".getBytes();
+
+   private static final byte OPEN_ITEM_CHAR = '{';
+   private static final byte SEPARATOR = ',';
+   private static final byte CLOSE_ITEM_CHAR = '}';
+   private final boolean includeMetadata;
+
+   public CacheEntryStreamProcessor(Publisher<CacheEntry<?, ?>> publisher, boolean keysAreJson, boolean valuesAreJson,
+         boolean includeMetadata) {
+      super(publisher);
+      this.keysAreJson = keysAreJson;
+      this.valuesAreJson = valuesAreJson;
+      this.includeMetadata = includeMetadata;
+   }
+
+   @Override
+   ByteBufSubscriber<CacheEntry<?, ?>> subscriber(ChannelHandlerContext ctx, ByteBufAllocator allocator) {
+      return new EntrySubscriber(ctx, allocator);
+   }
+
+   class EntrySubscriber extends ByteBufSubscriber<CacheEntry<?, ?>> {
+
+      protected EntrySubscriber(ChannelHandlerContext ctx, ByteBufAllocator allocator) {
+         super(ctx, allocator);
+      }
+
+      @Override
+      ByteBuf writeItem(CacheEntry<?, ?> item, ByteBuf pending) {
+         assert pending.writableBytes() > 0;
```

Comment:
This assertion might fail in a very specific scenario.

After writing an entry to the `pending` (or the newly created buffer) and it has only one writable byte, this method returns, then we receive a new item on `onNext` and write `,` to the buffer, which will be full when we try to write the entry. I guess just increasing the `requiredBytes` by one at line 75 might solve this.

Reply:
Good catch. I need to rethink this slightly.